src/ResourceManager/Common/Commands.Common.Strategies/AsyncOperationVisitor.cs

```diff
@@ -0,0 +1,55 @@
+﻿using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Commands.Common.Strategies
+{
+    /// <summary>
+    /// It's a base class for asynchronous operation visitors, such as 
+    /// CreateOrUpdateAsyncOperation and GetAsyncOperation visitors.
+    /// </summary>
+    abstract class AsyncOperationVisitor : IResourceBaseConfigVisitor<Task<object>>
+    {
+        public IClient Client { get; }
+
+        public CancellationToken CancellationToken { get; }
+
+        public State Result { get; } = new State();
+
+        public async Task<object> GetOrAddUntyped(IResourceBaseConfig config)
+            => await TaskMap.GetOrAdd(
+                config.DefaultIdStr(),
+                async _ =>
+                {
+                    var model = await config.Apply(this);
+                    if (model != null)
+                    {
+                        Result.GetOrAddUntyped(config, () => model);
+                    }
+                    return model;
+                });
+
+        public async Task<Model> GetOrAdd<Model>(IResourceBaseConfig<Model> config)
```

Comment:
This should be `TModel` to denote that it is a generic.

Reply:
Done.